Add PreferredAlias into OriginalNameAttribute

anandolee opened the pull request:
There might be duplicated enum values when allow_alias is true. Add Distinct() for Enum Dictionary used in JsonFormatter to remove the duplication.

fix #2405

bazel-io commented:
Can one of the admins verify this patch?

anandolee commented:
Jon, do you know if Distinct() keeps the first item for duplication?

jskeet commented:
It keeps the first one it encounters, but there's no guarantee which one will be returned first by reflection, so it's unstable, basically.

An alternative option would be to make protoc emit an attribute for fields to ignore when building the dictionary. That would allow us to make the decision in a consistent way, but would require regenerating the C# code before it would be effective, of course.

We could keep the current approach *and* add the attribute, so that we'd never throw an error, but the results might be inconsistent until code was regenerated - what do you think about that solution?

anandolee commented on csharp/src/Google.Protobuf.Test/TestProtos/UnittestProto3.cs:
```diff
@@ -209,8 +209,8 @@ public enum TestEnumWithDupValue {
     [pbr::OriginalName("BAR1")] Bar1 = 2,
     [pbr::OriginalName("BAZ")] Baz = 3,
-    [pbr::OriginalName("FOO2")] Foo2 = 1,
-    [pbr::OriginalName("BAR2")] Bar2 = 2,
```
Does Empty the OriginalNameAttribute for duplicate enum values make sense to you? So there's no need to add more attribute. We do not need the OriginalName for duplicate enum.

If so, I will go head change the other cods like protoc.

anandolee commented on csharp/src/Google.Protobuf.Test/TestProtos/UnittestProto3.cs:
```diff
@@ -209,8 +209,8 @@ public enum TestEnumWithDupValue {
-    [pbr::OriginalName("BAR2")] Bar2 = 2,
+    [pbr::OriginalName("")] Foo2 = 1,
+    [pbr::OriginalName("")] Bar2 = 2,
   }
```
Does Empty the OriginalNameAttribute for duplicate enum values make sense to you? So there's no need to add more attribute. We do not need the OriginalName for duplicate enum.

If so, I will go ahead change the rest code and protoc.

jskeet replied:
Rather than losing the information (even if we're not using it at the moment) I think it would be better to add an extra `bool` property to the attribute, e.g.

```csharp
[pbr::OriginalName("FOO1")] Foo1 = 1,
[pbr::OriginalName("BAR1")] Bar1 = 2,
[pbr::OriginalName("FOO2", PreferredAlias = false )] Foo2 = 1,
[pbr::OriginalName("BAR2", PreferredAlias = false)] Bar2 = 2,
```

It feels cleaner in terms of not trying to make one bit of information represent two things. (It may also be useful in terms of people reading the code.)

anandolee replied:
Have added the PreferredAlias into OriginalName

jskeet commented on csharp/src/Google.Protobuf/Reflection/OriginalNameAttribute.cs:
```diff
@@ -47,12 +47,19 @@ public class OriginalNameAttribute : Attribute
         public string Name { get; set; }
 
         /// <summary>
+        /// If the name is preffered in the .proto file.
```
Nit: preffered => preferred

anandolee commented:
Thanks for review!
I think it's fine not deal with the already-generated protos. Squashed into one commit and added the change for codegen. Will submit after travis green
